syntax = "proto3";

package bookstore;

import "shared/order.proto";

import "google/protobuf/empty.proto";

service OrderQueueService {
    rpc EnqueueOrder (QueueOrderRequest) returns (ErrorResponse);
    rpc DequeueOrder (google.protobuf.Empty) returns (DequeueOrderResponse);
    rpc CoordinateExecutors (CoordinateRequest) returns (CoordinateResponse);
    rpc NewElection(google.protobuf.Empty) returns (google.protobuf.Empty);
}

message QueueOrderRequest {
    ExecInfo info = 1;
    repeated OrderedBook booksInCart = 2;
    string name = 3;
    CreditCard CreditCard = 4;
    BillingAddress BillingAddress = 5;
}

message DequeueOrderResponse {
    oneof result {
      QueueOrderRequest order = 1;
      ErrorResponse error = 2;
    }
}

message OrderedBook {
    string title = 1;
    int64 quantity = 2;
}

message CoordinateRequest{
    string portnumber = 1;
}

message CoordinateResponse{
    repeated string ids = 1;
}